File: src/YowThi.Erp.Web/src/features/outsourced/outsourcedWorkspace.ts
import { getApiJson } from '../../app/api/apiTransport';
import type { OperationalLocale } from '../../app/i18n/locale';

export interface OutsourcedWorkspaceListItem {
  id: string;
  supplyDate: string;
  outsourcedVendorId: string;
  outsourcedVendorDisplayName: string;
  lifecycleStatus: string;
  rowVersion: number;
  createdAt: string;
  closedAt: string | null;
  deletedAt: string | null;
}

export interface OutsourcedWorkspaceDetail {
  id: string;
  salesProductId: string;
  salesProductDisplayName: string;
  quantity: number;
  pricingBasis: string;
  unitPrice: number;
  amountThb: number;
  rowVersion: number;
  recordedAt: string;
  deletedAt: string | null;
}

export interface OutsourcedWorkspace {
  id: string;
  supplyDate: string;
  outsourcedVendorId: string;
  outsourcedVendorDisplayName: string;
  lifecycleStatus: string;
  rowVersion: number;
  createdAt: string;
  closedAt: string | null;
  deletedAt: string | null;
  details: OutsourcedWorkspaceDetail[];
}

interface ListQuery {
  locale: OperationalLocale;
  search?: string;
  offset?: number;
  limit?: number;
  signal?: AbortSignal;
}

interface ListResponse {
  items: OutsourcedWorkspaceListItem[];
  nextOffset: number | null;
}

export function listOutsourcedWorkspace(query: ListQuery): Promise<ListResponse> {
  const parameters = new URLSearchParams();
  const search = query.search?.trim();
  if (search) parameters.set('search', search);
  if (query.offset !== undefined) parameters.set('offset', String(query.offset));
  if (query.limit !== undefined) parameters.set('limit', String(query.limit));
  const suffix = parameters.size === 0 ? '' : `?${parameters.toString()}`;
  return getApiJson<ListResponse>(`/api/v1/outsourced/workspace/${suffix}`, {
    locale: query.locale,
    signal: query.signal,
  });
}

export function getOutsourcedWorkspace(
  outsourcedSupplyBatchId: string,
  locale: OperationalLocale,
  signal?: AbortSignal,
): Promise<OutsourcedWorkspace> {
  return getApiJson<OutsourcedWorkspace>(
    `/api/v1/outsourced/workspace/${encodeURIComponent(outsourcedSupplyBatchId)}`,
    { locale, signal },
  );
}

File: src/YowThi.Erp.Application/Outsourced/OutsourcedWorkspace.cs
namespace YowThi.Erp.Application.Outsourced;

public sealed record OutsourcedWorkspaceListQuery(
    string Locale,
    string? Search,
    int Offset,
    int Limit);

public sealed record OutsourcedWorkspaceListItem(
    Guid Id,
    DateOnly SupplyDate,
    Guid OutsourcedVendorId,
    string OutsourcedVendorDisplayName,
    string LifecycleStatus,
    long RowVersion,
    DateTimeOffset CreatedAt,
    DateTimeOffset? ClosedAt,
    DateTimeOffset? DeletedAt);

public sealed record OutsourcedWorkspaceListPage(
    IReadOnlyList<OutsourcedWorkspaceListItem> Items,
    int? NextOffset);

public sealed record OutsourcedWorkspaceDetailItem(
    Guid Id,
    Guid SalesProductId,
    string SalesProductDisplayName,
    decimal Quantity,
    string PricingBasis,
    decimal UnitPrice,
    long AmountThb,
    Guid ReceiptStorageLocationId,
    string ReceiptStorageLocationDisplayName,
    long RowVersion,
    DateTimeOffset RecordedAt,
    DateTimeOffset? DeletedAt);

public sealed record OutsourcedWorkspace(
    Guid Id,
    DateOnly SupplyDate,
    Guid OutsourcedVendorId,
    string OutsourcedVendorDisplayName,
    string LifecycleStatus,
    long RowVersion,
    DateTimeOffset CreatedAt,
    DateTimeOffset? ClosedAt,
    DateTimeOffset? DeletedAt,
    IReadOnlyList<OutsourcedWorkspaceDetailItem> Details);

public interface IOutsourcedWorkspaceReader
{
    ValueTask<OutsourcedWorkspaceListPage> GetBatchesAsync(
        OutsourcedWorkspaceListQuery query,
        CancellationToken cancellationToken);

    ValueTask<OutsourcedWorkspace?> GetBatchAsync(
        Guid outsourcedSupplyBatchId,
        string locale,
        CancellationToken cancellationToken);
}
